NSAutoreleasePool instance should be added to all threads.

AaronRobinsonMSFT opened the pull request:
This expands the current support where support was only added to `ThreadPool` threads. Now all managed threads [~~or new native threads entering the runtime~~](https://github.com/dotnet/runtime/pull/52023#discussion_r627076254) will get an `NSAutoreleasePool` allocated.

New feature switch, "System.Threading.Thread.EnableAutoreleasePool", was created and the "System.Threading.ThreadPool.EnableDispatchAutoreleasePool" was removed.

Updated AutoReleaseTest for the new scenarios. Also added support for tests to pass [Runtime Configuration Options](https://docs.microsoft.com/dotnet/core/run-time-config/) to `corerun` - uses the syntax found [here](https://github.com/dotnet/runtime/blob/main/docs/workflow/trimming/feature-switches.md) - @LakshanF, @BruceForstall.

This is a follow up PR started with https://github.com/dotnet/runtime/pull/47592 

Contributes to https://github.com/xamarin/xamarin-macios/issues/11256

/cc @rolfbjarne @jkotas @jkoritzinsky @elinor-fung

AaronRobinsonMSFT committed:
Platform scripts now respect the MSBuild RuntimeHostConfigurationOption
 item and pass it to corerun.

AaronRobinsonMSFT committed:
Add an NSAutoreleasePool to all runtime related threads.

This expands the current support where support was only added to
ThreadPool threads. Now all managed threads or new native threads
entering the runtime will get an NSAutoreleasePool allocated.

New feature switch was created and the ThreadPool one was removed.
 - System.Threading.Thread.EnableAutoreleasePool

Updated AutoReleaseTest for the new scenarios.

jkoritzinsky commented on src/libraries/Native/Unix/System.Native/pal_autoreleasepool.m:
```diff
@@ -42,3 +42,9 @@ void SystemNative_DrainAutoreleasePool(void* pool)
+void SystemNative_AllocateThreadlocalAutoreleasePool(void)
+{
+    EnsureNSThreadIsMultiThreaded();
+    (void)SystemNative_CreateAutoreleasePool();
```
Is the created NSAutoreleasePool object guaranteed to be drained at thread exit?

AaronRobinsonMSFT replied:
Yes. From https://developer.apple.com/documentation/foundation/nsautoreleasepool#1651513.

>Each thread (including the main thread) maintains its own stack of NSAutoreleasePool objects (see Threads). As new pools are created, they get added to the top of the stack. When pools are deallocated, they are removed from the stack. Autoreleased objects are placed into the top autorelease pool for the current thread. When a thread terminates, it automatically drains all of the autorelease pools associated with itself.

AaronRobinsonMSFT committed:
Build issues.

jkotas commented on src/libraries/Native/Unix/System.Native/pal_autoreleasepool.m:
```diff
@@ -42,3 +42,9 @@ void SystemNative_DrainAutoreleasePool(void* pool)
     [((NSAutoreleasePool*)pool) drain];
 }
+
+void SystemNative_AllocateThreadlocalAutoreleasePool(void)
```
This does the same thing as `SystemNative_CreateAutoreleasePool`. Why do we need two different methods?

AaronRobinsonMSFT replied:
No real reason. This was created when I thought I needed to do some TLS related operations. I will remove this one.

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Threading/Thread.OSX.cs:
```diff
@@ -0,0 +1,39 @@
+    {
+        private static bool CheckEnableAutoreleasePool()
+        {
+#if !CORECLR
```
Why this ifdef? This switch was enabled for threadpool for all runtimes before this change.

AaronRobinsonMSFT replied:
This is an implementation detail that needs to be decided. My starting premise here is `ThreadPool` threads shouldn't be special cased in any runtime. In this case the type originally used, `AppContextConfigHelper`, works fine but only after the implementing runtime has populated it. This works with `ThreadPool` because it wasn't needed so early. In order to read the needed setting earlier enough in CoreCLR, we need to use the `CLRConfig` type. I can fallback to `AppContextConfigHelper` for the non-CoreCLR scenario but that technically work as expected until Mono respects the feature for all threads. At present it would only work for `ThreadPool` threads for all runtimes and not all threads so I think this behavior is the correct one until that work happens in Mono.

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Threading/Thread.OSX.cs:
```diff
@@ -0,0 +1,39 @@
+        internal static void AllocateThreadlocalAutoreleasePool()
+        {
+            if (!ThreadOSX.EnableAutoreleasePool)
+                throw new PlatformNotSupportedException();
```
Should this be no-op instead of PNSE when the switch is not set? It is the behavior for threadpool.

AaronRobinsonMSFT replied:
I am hesitant to do this. I think knowing this is enabled as expected is far better than realizing much later that an `NSAutoreleasePool` hasn't been allocated. More than willing to make it a no-op but I'm reluctant to make these sorts of policy decisions no-ops when called.

AaronRobinsonMSFT replied:
Actually I think I get your suggestion now. I will make this a no-op.

AaronRobinsonMSFT committed:
Remove unnecessary new native export for allocating an NSAutoreleasePool.
If feature is not enabled, allocating a NSAutoreleasePool is a no-op.

AaronRobinsonMSFT committed:
Update header.

AaronRobinsonMSFT commented:
@lambdageek I took a quick look at how to call these P/Invokes in mono. It isn't obvious to me how to try adding support. My best guess is around 

https://github.com/dotnet/runtime/blob/01b7e73cd378145264a7cb7a09365b41ed42b240/src/mono/mono/metadata/threads.c#L1160-L1173

If you think this support should be easy to add, I'm willing to take a few hours to try it out.

lambdageek commented:
> @lambdageek I took a quick look at how to call these P/Invokes in mono. It isn't obvious to me how to try adding support. My best guess is around
> 
> https://github.com/dotnet/runtime/blob/01b7e73cd378145264a7cb7a09365b41ed42b240/src/mono/mono/metadata/threads.c#L1160-L1173
> 
> If you think this support should be easy to add, I'm willing to take a few hours to try it out.

What needs to happen?  a call to `Thread.AllocateThreadlocalAutoreleasePool` ?

yea I think the right place is right before **both** callers of `fire_attach_profiler_events`.  the spot you found is one, the other is 

https://github.com/dotnet/runtime/blob/01b7e73cd378145264a7cb7a09365b41ed42b240/src/mono/mono/metadata/threads.c#L1600-L1601

That one is would set up an autorelease pool when a foreign thread attaches to the runtime via a pinvoke callback.   (eg: managed code calls a pinvoke and passes it a delegate (or an `UnmanagedCallersOnly` function pointer), the native code creates a new thread and calls the unmanaged function pointer from the new thread).

In general the pattern to follow is like this: 

https://github.com/dotnet/runtime/blob/01b7e73cd378145264a7cb7a09365b41ed42b240/src/mono/mono/metadata/threads.c#L878-L898

(some things to note: 1. if the method may not be present (removed by the IL Linker, or not defined on every plaform), don't assert; 2. `mono_runtime_get_no_exec` (check if we're the AOT compiler) is probably needed - a little bit of EE initialization happens even just during AOT compilation, and some of this thread attach code probably runs. 3. i'm assuming we want to swallow any managed exceptions, so the final mono_error_cleanup will swallow them. 4. since the method takes no arguments, the `HANDLE_FUNCTION_ENTER`/`EXIT` business is not needed and the actual call is more like `mono_runtime_invoke_checked (method, NULL, NULL, error)` )

AaronRobinsonMSFT committed:
Initialize for platform prior to transferring control.

AaronRobinsonMSFT committed:
Address Linker issues by defining a throwing version of the API the
linker can keep even though the runtime won't call the function.

AaronRobinsonMSFT committed:
Update files for building based on their names.

AaronRobinsonMSFT committed:
TargetsUnix is also set for OSX - sigh.

jkotas commented on src/coreclr/vm/threads.cpp:
```diff
@@ -1562,7 +1588,10 @@ Thread::Thread()
     m_monitorLockContentionCount = 0;
 
-    InitContext();
+    {
+        GCX_COOP_NO_THREAD_BROKEN();
+        m_pDomain = SystemDomain::System()->DefaultDomain();
+    }
```
```suggestion
    m_pDomain = SystemDomain::System()->DefaultDomain();
```
This `GCX_COOP_NO_THREAD_BROKEN` should not be needed. Everything in `SystemDomain::System()->DefaultDomain()` are `LIMITED_METHOD_DAC_CONTRACT` methods (ie they don't do anything interesting)

AaronRobinsonMSFT replied:
Will do.

jkotas commented on src/tests/Interop/ObjectiveC/autorelease.mm:
```diff
@@ -31,3 +32,10 @@ - (void)release
+using ManagedCallback = void(*)(AutoReleaseTest*);
+extern "C" DLL_EXPORT void passAndCallOnNativeThread(AutoReleaseTest* art, ManagedCallback callback)
+{
+    std::thread th{ callback, art };
```
I understand that this does what we agreed to, but I am wondering what it means for reliability.

The way we create the autorelease pool for the external threads violates Apple guidable: You should always drain an autorelease pool in the same context (invocation of a method or function, or body of a loop) that it was created. (From https://developer.apple.com/documentation/foundation/nsautoreleasepool.)

What happens when there is autorelease pool created on the thread already, and we push the abandoned people on top of it? i.e.

ExternalThread()
{
    3rd party ObjectiveC code

    @autoreleasepool {
        .... the 3rd party ObjectiveC code here will call into the .NET runtime that will push our new abandoned autorelease pool ...
    }

    more 3rd party Objective code
}
```

AaronRobinsonMSFT replied:
> The way we create the autorelease pool for the external threads violates Apple guideline: You should always drain an autorelease pool in the same context (invocation of a method or function, or body of a loop) that it was created.

Agreed. Xamarin does basically what we do here minus the implicit drain. The release is only when it detaches from the runtime - at least that is how I understand the Mono Embedding APIs work.

> What happens when there is autorelease pool created on the thread already, and we push the abandoned people on top of it?

My understanding is the above `autoreleasepool` block will add a new pool prior to calling into the runtime. When it exits it would pop the one we would have added off the top of the thread's pool stack. This would then cause a pool that doesn't drain. I'm not sure how we want to handle this. Perhaps @rolfbjarne has an opinion here. I can move this around without a lot of issue.

rolfbjarne replied:
> > The way we create the autorelease pool for the external threads violates Apple guideline: You should always drain an autorelease pool in the same context (invocation of a method or function, or body of a loop) that it was created.
> 
> Agreed. Xamarin does basically what we do here minus the implicit drain. The release is only when it detaches from the runtime - at least that is how I understand the Mono Embedding APIs work.
> 
> > What happens when there is autorelease pool created on the thread already, and we push the abandoned people on top of it?
> 
> My understanding is the above `autoreleasepool` block will add a new pool prior to calling into the runtime. When it exits it would pop the one we would have added off the top of the thread's pool stack. This would then cause a pool that doesn't drain. I'm not sure how we want to handle this. Perhaps @rolfbjarne has an opinion here. I can move this around without a lot of issue.

I've always wanted Xamarin to follow Apple's guidelines, but it wasn't my decision, nor did our approach ever break, so I never pushed for it (and there's always broken stuff that ends up taking priority).

In my experience Apple tends to break code that doesn't follow their guidelines, so I've been kind of waiting for this to break at some point.

My opinion is that even if the current approach works today, it's preferable to follow the guidelines.

AaronRobinsonMSFT replied:
>My opinion is that even if the current approach works today, it's preferable to follow the guidelines.

@rolfbjarne Sounds good.

The plan is therefore to update this PR to the following:

1) Newly created managed threads will get a `NSAutoreleasePool` allocated for them and then during termination of the thread we will explicitly call `drain` on the pool. We will not be relying on the statement in [doc](https://developer.apple.com/documentation/foundation/nsautoreleasepool) that states: "When a thread terminates, it automatically drains all of the autorelease pools associated with itself."

1) When a thread enters the runtime from unmanaged code a new `NSAutoreleasePool` will be allocated for the thread. When the thread returns from the Reverse P/Invoke we will call `drain` on the pool. This does mean there may be some impact to scenarios such as methods marked with `UnmanagedCallersOnly`. Optimizations could exist here to skip allocating/draining the pool when a thread that already has an runtime allocated pool enters from unmanaged code.

AaronRobinsonMSFT replied:
I think we are going to augment the above plan of record. Step (1) is still a go, however (2) is not. The current plan is to _not_ push an `NSAutoreleasePool` when an external thread enters the runtime for the first time. I've documented our reasoning in code and will duplicate it here. In this case, the term "platform initialization" indicates pushing an `NSAutoreleasePool`.

```c++
// When a new thread enters the runtime through a Reverse P/Invoke
// skipping platform initialization is the preferred manner. The reasoning
// is based on the fact that users may have already initialized this thread
// for the platform in question. Attempting to re-initialize this thread is
// therefore imposing policy in a manner that may be undesirable.
```

AaronRobinsonMSFT committed:
Update src/coreclr/vm/threads.cpp

Co-authored-by: Jan Kotas <jkotas@microsoft.com>

AaronRobinsonMSFT commented:
The current test failures all appear to be related to Tiered Compilation. As soon as I set `DOTNET_TieredCompilation=0` the tests pass locally &ndash; if I don't they repro with the hang. I did some investigation and determined that calling into the managed callout I added so early in thread initialization appears to be the culprit. It destabilizes some invariant in the Tiered Compilation system and causes hangs. To further validate this I removed the call out to managed code and all the tests past locally with and without Tiered Compilation. I've reached out to the Tiered Compilation owners offline for help investigating this.

Next up I will be try calling into managed and immediately returning instead of calling back into the P/Invoke to allocate the `NSAutoreleasePool`.

AaronRobinsonMSFT committed:
Handle other cases where specific platform initialization isn't
needed or possible in some cases (e.g. GC).

jkotas commented on src/coreclr/vm/threads.h:
```diff
@@ -579,9 +579,25 @@ enum ThreadpoolThreadType
+{
+    SUTF_None = 0,
+
+    // When the thread enters the thread store, the lock should be taken.
```
What does it mean for a thread to enter the thread store?

This flag and how it is used does not make a whole lot of sense to me.

jkotas replied:
> What does it mean for a thread to enter the thread store?

Ok, I got confused by some of the places where this is used. I see that it is used in AddThread that the comment is referring to.

Why is it safe to ship the thread store locks for some threads? I understand that it is safe to skip thread store locks early during startup because there is nothing else running at that point. But I do not think it is ever safe to skip the thread store lock to add new threads once the runtime is started up.

AaronRobinsonMSFT replied:
> What does it mean for a thread to enter the thread store?

The `ThreadStore` class seems to be the keeper of `Thread` instances. Based on some cursory comments it appears to be used to be a singular place to enumerate them. Previously this flag was a `BOOL` passed in. I've not constructed a complete mental model of this code yet so don't have a better answer for why unfortunately. I can say that the only threads that don't currently take this lock are GC threads so I might be able to elide it entirely. It would likely become a specialized `SetupUnstartedThread` instance (e.g., `SetupUnstartedGCThread`).

AaronRobinsonMSFT replied:
> But I do not think it is ever safe to skip the thread store lock to add new threads once the runtime is started up.

That is true. This flag is only used in places where a lock may be taken - in `ThreadStore::AddThread()` and `Thread::SetBackground()`. Both these functions had a default value of `TRUE` and that was only ever set by the GC  early in a threads life. Based on your statement I am going to say this shouldn't be a high level concept and will instead hide it behind some GC specific idea and have it "just work" for all other scenarios.

AaronRobinsonMSFT committed:
Remove the high level ThreadStoreLock flag and instead associate it
with the thread type needed - GC.

jkotas commented on src/coreclr/vm/threads.h:
```diff
@@ -1157,6 +1171,8 @@ class Thread
 #ifdef FEATURE_COMINTEROP_APARTMENT_SUPPORT
         TT_CallCoInitialize       = 0x00000002, // CoInitialize needs to be called.
 #endif // FEATURE_COMINTEROP_APARTMENT_SUPPORT
+        TT_SkipThreadStoreLock    = 0x00000004,    // Skip thread store lock. GC is only consumer.
```
Unused

jkotas commented on src/coreclr/vm/threads.h:
```diff
@@ -1296,6 +1312,15 @@ class Thread
         return HasThreadStateNC(Thread::TSNC_EtwStackWalkInProgress);
     }
 
+    BOOL RequireThreadStoreLock()
```
I have looked more into where this is used: The background GC threads that will have it set can be created and destroyes while the program is running. It is still not clear to me what makes it safe to skip the threadstore lock for them. What prevents some other thread from being added to `s_pThreadStore->m_ThreadList` at the same time the GC thread is being added to `s_pThreadStore->m_ThreadList`?

(It is possible that the existing logic had a latent bug that does not happen often enough to get noticed.)

AaronRobinsonMSFT replied:
>It is still not clear to me what makes it safe to skip the threadstore lock for them. What prevents some other thread from being added to s_pThreadStore->m_ThreadList at the same time the GC thread is being added to s_pThreadStore->m_ThreadList?

Right. I am still catching up but even the 30 minutes I took last night confirmed there is a fair bit of complexity in this area. The implementation of the lock/unlock calls out to the `ThreadSuspend` class and I've never used this class at all. I'm going to need to take a few hours to debug/read in order to answer this question.

This is what I get for touching things 😄

AaronRobinsonMSFT replied:
@jkotas I agree this name is bad because there are places where the BGC thread can get into trouble so I am going to rename it entirely. Note that only the BGC has this logic so lets focus on how that scenario works. At present, there are 3 times when the new `RequireThreadStoreLock()` API is called.

1) `ThreadStore::AddThread()`
1) `Thread::SetBackground()`
1) `ThreadStore::TransferStartedThread()`

Scenarios (1) and (2) are fine because they are called in a gen2 GC when the BGC is first launched. This, based on my understanding means that no other threads are running so it is okay to do this. Cases when there could be other threads, I believe Server GC, are guarded because this create call is done within a `join` clause in the GC and with synchronization around the `prepare_bgc_thread()` method.

Scenario (3) is a bit more of an issue. This is called by the newly created BGC thread which means it can be running with other managed threads and as pointed out we could get into trouble. The beginning of this is in `HasStarted()`.

I am inclined to update the API name, conditionally take the lock for (1) and (2) based on BGC, and always take the lock in (3).

AaronRobinsonMSFT replied:
Stand by. I just tried to let (3) take the lock and now I've got a deadlock. Investigating...

jkotas replied:
> This, based on my current understanding, means that no other threads are running so it is okay to do this

It means that there are no threads running in cooperative mode. There can be other threads running in preemptive mode.

> conditionally take the lock for (1) and (2) based on BGC, and always take the lock in (3).

I guess we avoid taking the threadstore lock in (1) and (2) because of it is taken already, and taking it a second time would cause asserts or crashes. If it is the case, would it make sense to call the argument `fThreadStoreLockAlreadyTaken` and then assert in the implementation that the thread store lock is really taken when this argument is true.

Also, the problem with SetBackground thread can be solved by creating all threads as background by default, and only call this API in the minority places that create foreground threads. The foreground threads are the outlier that needs an extra work. Background threads are the common case.

AaronRobinsonMSFT replied:
Doh. The (3) is actually for the same reason as (1) and (2). The creator thread is waiting for the BGC to get to a certain point before continuing. This effectively lets the BGC safely start up without the lock and then never needs it again. Basically what we have here is correct so no bug - yay. However, the entire naming situation here is confusing. I will take the `fThreadStoreLockAlreadyTaken` suggestion because that is the correct wording in this case.

After that I will look into the `SetBackground()` suggestion. That appears more complicated given the number of places that is called and not called so will require some more auditing.

AaronRobinsonMSFT replied:
@jkotas I've applied the "already taken" concept and this makes things much clearer. Moving onto the `SetBackground()` suggestion.

AaronRobinsonMSFT replied:
Deferring any additional cleanup to another PR.

marek-safar commented on docs/workflow/trimming/feature-switches.md:
```diff
@@ -17,7 +17,7 @@ configurations but their defaults might vary as any SDK can set the defaults dif
 | HttpNativeHandler | System.Net.Http.UseNativeHttpHandler | HttpClient uses by default platform native implementation of HttpMessageHandler if set to true. |
 | StartupHookSupport | System.StartupHookProvider.IsSupported | Startup hooks are disabled when set to false. Startup hook related functionality can be trimmed. |
-| TBD | System.Threading.ThreadPool.EnableDispatchAutoreleasePool | When set to true, creates an NSAutoreleasePool around each thread pool work item on applicable platforms. |
+| TBD | System.Threading.Thread.EnableAutoreleasePool | When set to true, creates an NSAutoreleasePool for each thread and thread pool work item on applicable platforms. |
```
Please update the substitution file as well https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/ILLink/ILLink.Substitutions.OSX.xml

AaronRobinsonMSFT replied:
@marek-safar That was done in this PR already.

AaronRobinsonMSFT committed:
Remove unused enum.
Remove GetUniqueEEId() logic.
Remove check for AbortRequest from TransferStartedThread().
  - This can be done because it is only called from HasStarted() so
    the user has no chance to make the request and we don't support
    custom hosting.
  - The only other place this work occurs is during a FuncEval - which
    again means the thread is running manage code and after the call to
    the HasStarted() call.

AaronRobinsonMSFT committed:
Rename m_Creater to m_Creator.
Create new TSNC flag for indicating the ThreadStoreLock is held
  during the start up of the thread - TSNC_TSLTakenForStartup.

AaronRobinsonMSFT commented on src/coreclr/vm/threads.cpp:
```diff
@@ -570,73 +570,22 @@ DWORD Thread::StartThread()
-// Creates Thread for current thread if not previously created.
-// Returns NULL for failure (usually due to out-of-memory.)
-//-------------------------------------------------------------------------
-Thread* SetupThreadNoThrow(HRESULT *pHR)
```
This was just moved down.

AaronRobinsonMSFT commented on src/coreclr/vm/threads.cpp:
```diff
@@ -1857,90 +1876,88 @@ BOOL Thread::HasStarted(BOOL bRequiresTSL)
     }
     EX_END_CATCH(SwallowAllExceptions);
 
-FAILURE:
```
I broke apart the `if`/`else` and made a always failure `goto`. The contents of the two branches are largely the same apart from the locking boolean in the failure case &ndash; which is now the `TSNC_TSLTakenForStartup` flag.

AaronRobinsonMSFT committed:
Merge remote-tracking branch 'upstream/main' into autoreleasepool_per_thread

AaronRobinsonMSFT committed:
Root the AllocateThreadlocalAutoreleasePool function.

AaronRobinsonMSFT committed:
The eventpipe streaming thread shouldn't init plaform.

AaronRobinsonMSFT committed:
Add explicit drain call during managed thread termination.

jkotas commented on src/libraries/System.Private.CoreLib/src/ILLink/ILLink.Descriptors.Shared.xml:
```diff
@@ -19,6 +19,9 @@
     <type fullname="System.Threading.Timer">
       <property name="AllTimers" />
     </type>
+    <type fullname="System.Threading.Thread">
```
Since this is a workaround for a CoreCLR specific issue, it would fit better into `src\coreclr\System.Private.CoreLib\src\ILLink\ILLink.Descriptors.xml`

jkotas commented on src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems:
```diff
@@ -2027,10 +2025,19 @@
       <Link>Interop\Windows\Kernel32\Interop.Threading.cs</Link>
     </Compile>
   </ItemGroup>
+  <ItemGroup Condition="'$(TargetsUnix)' == 'true' or '$(TargetsBrowser)' == 'true'">
```
I am not sure whether it is an improvement to move the ifdefs from here to the .cs file, for one of the low-level threading related files. (The actual problem is that the threading in CoreCLR is waiting to be refactored into managed. Once that happens, the `FeatureCoreCLR` condition will disapper around all threading related files here.)

AaronRobinsonMSFT replied:
It is an improvement only because I need a place to keep a method from being complained about due to linker nonsense. I've been struggling for the past 2 days almost entirely trying to get the linker to not strip the function or stop complaining that it isn't there. It was suggested that I get it to work so I'm not blocked and move on. The biggest issue right now from my understanding is https://github.com/dotnet/runtime/issues/51991.

jkotas replied:
#51991 is CoreCLR specific problem, so it would be nice to keep the ifdefs and dummy methods required to workaround it under src/coreclr.

AaronRobinsonMSFT replied:
I can do that, but I would then like to rename these files then because they are not `Thread.Windows.cs` - they are `Thread.Mono.Windows.cs`. The issue here is the various levels and slices of the build. Some are in the project files some are in the CS files and then we have a mix of naming conventions. My expectation here is any file in the shared directory should be consumable by all runtimes. If we need to pare the file down then we should do that but not at the file level.

jkotas replied:
Most of the FeatureCoreCLR conditions in System.Private.CoreLib.Shared.projitems, including this one, are workarounds for https://github.com/dotnet/runtime/issues/31721 . 

We have been naming the files with assumption that this issue is fixed and CoreCLR is switched to use the shared implementation. Otherwise, the naming gets really hard. For example, some of the files would have to be WindowsOrWindowsEmulatorOnUnix to capture how they are used currently, etc.

AaronRobinsonMSFT replied:
Okay. I'll try something else.

AaronRobinsonMSFT committed:
Move rooting of CreateAutoreleasePool to coreclr XML files.
Remove platform initialization from new threads coming via Reverse P/Invoke.

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Threading/Thread.OSX.cs:
```diff
@@ -0,0 +1,43 @@
+    public sealed partial class Thread
+    {
+        [UnmanagedCallersOnly]
+        private static void CallDrain(IntPtr p)
```
This also got pretty CoreCLR specific. I would move it under coreclr/src. For other runtimes that have more of the threading refactored to managed code, it would be a regular managed call (e.g. from here - https://github.com/dotnet/runtime/blob/65bcecebde4aebd5926952ccfc4dc8e7fbf7eaed/src/mono/System.Private.CoreLib/src/System/Threading/Thread.Mono.cs#L305). No unmanaged pointers involved.

AaronRobinsonMSFT replied:
My intent here is to have Mono call this same API though for adding pools to managed threads.

AaronRobinsonMSFT replied:
I moved the API and will add something simpler to Mono. The exit point above makes sense and will hopefully make the Mono changes simpler.

AaronRobinsonMSFT committed:
Revert "Update files for building based on their names."

This reverts commit 543ed7affb2ba30e3dbb124b005a6ca7df0c7059.

AaronRobinsonMSFT committed:
rough_revert

AaronRobinsonMSFT committed:
Move CreateAutoreleasePool to CoreCLR specific cs file.

AaronRobinsonMSFT committed:
CreateAutoreleasePool should always exist on coreclr.

AaronRobinsonMSFT committed:
Create new FEATURE_ define for NSAutoreleasePool support in CoreCLR.

AaronRobinsonMSFT committed:
Redesign to make code shareable with Mono.

AaronRobinsonMSFT commented:
@lambdageek I've hit an issue with a race between the main thread and finalizer in mono. The issue is how we read in the feature settings &ndash; the managed `AppContext` is populated after the call we use to create an `NSAutoreleasePool`. This is fine for the main thread and regular managed threads when the runtime is fully up but for the finalizer is calls into managed code prior to the main thread populating the `AppContext`. The issue here is the global static is set too early. We had the same issue in CoreCLR. You can see this in the this PR &ndash; look for `SetupMainThread()` in `threads.cpp`. Any suggestions are welcome. I can also push up what I have behind an `#if 0` and let you all handle this at a future date.

AaronRobinsonMSFT committed:
Remove unnecessary rooting in XML.

AaronRobinsonMSFT committed:
Conditionally clean up from platform initialization.

lambdageek commented:
@AaronRobinsonMSFT either `#if 0` or a separate branch/draft PR is fine. I _think_ the right thing to do is to add a new `MONO_THREAD_CREATE_FLAGS_DEFER_PLATFORM_INITIALIZE` flag to `MonoThreadCreateFlags` and to thread that through from `mono_thread_create_internal` to `start_wrapper_internal`, if you want to give it a try.  If not, just put what you have some place where we can grab it.

---

@rolfbjarne so if I followed this thread - we aren't going to be setting up an NSAutoReleasePool for external threads that attach via a reverse P/Invoke?  I'm not sure if that's easy to do using the existing mono profiler event-based approach.  Should we try to do something here, or is it ok to leave it alone for now until Mono incorporates Aaron's code?

AaronRobinsonMSFT commented:
>I think the right thing to do is to add a new MONO_THREAD_CREATE_FLAGS_DEFER_PLATFORM_INITIALIZE flag to MonoThreadCreateFlags and to thread that through from the call to mono_thread_create_internal in init_finalizer_thread down into start_wrapper_internal, if you want to give it a try. If not, just put what you have some place where we can grab it.

@lambdageek W00t! That is exactly what I wanted to do but didn't think it was a good idea. I will give it a shot.

AaronRobinsonMSFT committed:
Add support in Mono

AaronRobinsonMSFT commented:
@lambdageek Added support in Mono.

lambdageek commented on src/mono/mono/metadata/threads.c:
```diff
@@ -3912,18 +3915,58 @@ mono_thread_init_apartment_state (void)
 	}
 #endif
+
+#if TARGET_OSX || HOST_MACCAT || TARGET_IOS || TARGET_TVOS
```
host/target confusion.

but also this can be simplified to 

```suggestion
#if defined(HOST_DARWIN)
```

(and in various other places)

lambdageek commented on src/mono/mono/metadata/threads.c:
```diff
@@ -3912,18 +3915,58 @@ mono_thread_init_apartment_state (void)
+
+	MONO_STATIC_POINTER_INIT_END (MonoMethod, create_autoreleasepool)
+
+	if (mono_runtime_get_no_exec ())
+		return;
```
I'm not 100% convinced the AOT compiler can get here, after all.  Let's drop it. The AOT compiler will break in CI if I'm wrong.

If it turns out this is necessary, I think it's slightly cleaner if the return is before the `create_autoreleasepool` initialization.

(Same thing for cleanup platform state).

lambdageek commented on src/mono/mono/metadata/threads.c:
```diff
@@ -3912,18 +3915,58 @@ mono_thread_init_apartment_state (void)
 {
-#ifdef HOST_WIN32
 	MonoInternalThread* thread = mono_thread_internal_current ();
+	if (!(thread->flags & MONO_THREAD_FLAG_PLATFORM_INIT))
+		return;
```
nit: `CoInitializeEx` is always called, but `CoUninitialize` is guarded by the flag.  They should be balanced.

AaronRobinsonMSFT replied:
I don't see how that is true &ndash; at least from `mono_thread_init_platform_state`. The last thing that `mono_thread_init_platform_state` does is flip this bit so when `mono_thread_cleanup_platform_state` is called it will run or simply return if it wasn't run.

lambdageek replied:
yea, you're right.  the early return in init is only on Darwin, and the CoInitializeEx is only on Windows.

lambdageek commented on src/mono/mono/metadata/threads.c:
```diff
@@ -3912,18 +3915,58 @@ mono_thread_init_apartment_state (void)
 	}
 #endif
+
+#if TARGET_OSX || HOST_MACCAT || TARGET_IOS || TARGET_TVOS
```
```suggestion
#if defined(HOST_DARWIN)
```

lambdageek reviewed (commented):
Mono bits LGTM, in general. Although I expected a deferred call to `mono_thread_init_platform_state` in the finalizer loop somewhere.  Perhaps after the semaphore wait.

A few minor suggestions, otherwise.

lambdageek commented on src/mono/mono/metadata/gc.c:
```diff
@@ -908,7 +908,7 @@ static void
 {
 	ERROR_DECL (error);
-	gc_thread = mono_thread_create_internal ((MonoThreadStart)finalizer_thread, NULL, MONO_THREAD_CREATE_FLAGS_NONE, error);
+	gc_thread = mono_thread_create_internal ((MonoThreadStart)finalizer_thread, NULL, MONO_THREAD_CREATE_FLAGS_DEFER_PLATFORM_INIT, error);
```
So should we call `mono_thread_init_platform_state` inside `finalizer_thread` somewhere?

AaronRobinsonMSFT replied:
Oops, good catch. Yes, the Finalizer thread should call that. I don't know about Mono but in CoreCLR the Finalizer thread lives the process lifetime so technically it will not drain until the process terminates.

lambdageek replied:
Yep, Mono's too.

AaronRobinsonMSFT replied:
Found a nasty bug adding this. I was using `mono_runtime_try_invoke_handle` instead of `mono_runtime_try_invoke_handle_void` and saw many asserts when I tried initializing the platform.

AaronRobinsonMSFT committed:
Review feedback for Mono.

rolfbjarne commented:
> @rolfbjarne so if I followed this thread - we aren't going to be setting up an NSAutoReleasePool for external threads that attach via a reverse P/Invoke? I'm not sure if that's easy to do using the existing mono profiler event-based approach. Should we try to do something here, or is it ok to leave it alone for now until Mono incorporates Aaron's code?

How does it work today with legacy Xamarin? Do the callbacks installed with `mono_profiler_install_thread` get called for external threads that attach via a reverse P/Invoke?

lambdageek commented:
> How does it work today with legacy Xamarin? Do the callbacks installed with `mono_profiler_install_thread` get called for external threads that attach via a reverse P/Invoke?

Yes.  the `thread_started` event is fired from `fire_attach_profiler_events` which is called from `mono_thread_internal_attach` which is called from `mono_threads_attach_coopl` which is called from native-to-managed wrappers when a reverse P/Invoke is called.

lambdageek reviewed (dismissed):
LGTM (Mono)

AaronRobinsonMSFT commented:
@jkotas @kouvel @jkoritzinsky @elinor-fung Mono's changes have been signed off and the CI is green. Any other CoreCLR feedback?

jkotas commented on src/coreclr/vm/threads.cpp:
```diff
@@ -852,13 +798,111 @@ Thread* SetupThread()
+Thread* SetupThreadNoThrow(HRESULT *pHR)
+{
+    WRAPPER_NO_CONTRACT;
+    return SetupThreadNoThrowWorker(pHR, /* performPlatformInit */ true);
```
Are all places that `SetupThreadNoThrow` called on threads created by the runtime? I think that I can see number of places where this is called for external threads.

I am wondering whether the logic to identify threads created by runtime (vs. external threads) should be more explicit. We should only needs this on these threads:
- Main thread
- Finalizer thread
- new Thread
- Threadpool threads

Is that right?

AaronRobinsonMSFT replied:
>Are all places that `SetupThreadNoThrow` called on threads created by the runtime? I think that I can see number of places where this is called for external threads.

There are several outstanding but those all appear to be on COM scenarios - which is long standing behavior. I am more than willing to handle that in a subsequent PR as there are other things I would like to clean up.

>I am wondering whether the logic to identify threads created by runtime (vs. external threads) should be more explicit.

That list is what I have in mind too. I can narrow that down more if desired but then we are getting into some COM related behavioral changes that I would like to avoid at present. I think for non-Windows platforms this is correct and for the Windows platform this is correct in spirit.

kouvel replied:
If it'll be made more explicit, I think the tiering and multi-core JIT background threads may also need this, as they may run some managed code while jitting.

jkotas replied:
Why not treat tiereing background thread the same way as external thread that wandered into the runtime?

kouvel replied:
I figure external threads are expected to provide their own pool, whereas these background threads are created by the runtime.

kouvel replied:
I see, in that case it would probably be ok to not have a pool for the background jitting threads

AaronRobinsonMSFT replied:
Is this a priority to fix in this PR or can it be deferred?

jkotas replied:
I think that the whole PR can be signficantly simplified with this plan.
- The runtime created threads can be handled by adding the callback to `Thread.StartHelper.RunWorker`. No unmanaged handshake required.
- The main thread can be handled by adding the callback to the same place as `StartupHookProvider.RunStartupHooks`
- The finalizer thread is the only one that needs a bit more complex handshake to delay the start

AaronRobinsonMSFT replied:
> The runtime created threads can be handled by adding the callback to Thread.StartHelper.RunWorker. No unmanaged handshake required.

Done.

> The main thread can be handled by adding the callback to the same place as StartupHookProvider.RunStartupHooks

That is possible but we still need to clean up the main thread from native. Given that I am inclined to keep it all in native for both runtimes since it is logically symmetrical now. All managed Threads are done in managed and the Main and Finalizer thread are both handled on the native side.

jkotas replied:
It is fine to call this from native. I was more concerned about moving this closer to the point where we call Main method, and not during EEStartup. 

We have hosting entrypoint that initializes the runtime, without setting up the program entrypoint or calling the Main method (ie it just calls EEStartup, and not the rest). I do not think that it makes sense for this hosting entrypoint to setup the autorelease pool. I think it would be more appropropriate to setup the autorelease pool only once we are actually about to call the Main method. 

It matches what we do for COM: EEStartup does not initialize the COM on the thread. COM is only initialized only once we are about to call the Main method.

jkotas commented on src/mono/System.Private.CoreLib/src/System/Threading/Thread.Mono.cs:
```diff
@@ -246,6 +246,10 @@ private void StartCore()
         [DynamicDependency(nameof(StartCallback))]
         [DynamicDependency(nameof(ThrowThreadStartException))]
+#if TARGET_OSX || TARGET_MACCATALYST || TARGET_IOS || TARGET_TVOS
+        [DynamicDependency(nameof(CreateAutoreleasePool))]
```
Is this dynamic dependency sufficient to keep these mehods around for all cases? If I have a trivial app that does not create any new threads, `StartInternal` should be unused, but these methods are still going to be called for Main thread. Is it a problem?

AaronRobinsonMSFT replied:
I believe this is sufficient given ti is called from within Mono and is defined in `icall-def-netcore.h`. I truthfully don't know enough about the Linker and Mono to understand if this is sufficient but defer to Linker and Mono experts here - @vitek-karas @agocke @marek-safar @lambdageek

lambdageek replied:
Yea, CreateAutoreleasePool and DrainAutoreleasePool may need to go in the linker descriptor. I'm not sure if there's a way to only add it for specific target platforms.  If there are no calls to Thread.StartInternal, nothing would keep it or its dynamic dependencies alive.

On some platforms (webassembly) icalls in `icall-def-netcore.h` may be linked out if nothing in managed code uses them (look for "IcallTableGenerator").  So the IL linker does not treat icalls specially, AFAIK.  Managed methods invoked from inside Mono are also not automatically preserved.

AaronRobinsonMSFT replied:
Okay. Will add this to the descriptor as suggested.

AaronRobinsonMSFT committed:
Move linker root options to XML

AaronRobinsonMSFT committed:
Move NSAutoreleasePool alloc to managed code for managed Threads.

AaronRobinsonMSFT committed:
Revert unnecessary changes.

jkotas commented on src/coreclr/vm/threads.cpp:
```diff
@@ -2895,16 +2939,18 @@ void Thread::OnThreadTerminate(BOOL holdingLock)
-    }
 #endif // FEATURE_COMINTEROP_APARTMENT_SUPPORT
 
+        Thread::CleanUpForManagedThreadInNative(this);
```
Is this still necessary to call this here? I do not think it will ever do anything useful. We take care of all situations of all situations that I can think of elsewhere.

AaronRobinsonMSFT commented on src/libraries/System.Private.CoreLib/src/System/Threading/Thread.cs:
```diff
@@ -63,19 +63,31 @@ private void RunWorker()
-                if (start is ThreadStart threadStart)
+#if FEATURE_NSAUTORELEASEPOOL
+                Thread.CreateAutoreleasePool();
+                try
```
I saw that. I wasn't sure how the try/catch was going to impact it. If an exception here is degenerate then I think we can simply remove it and add a comment as to why it is missing.

AaronRobinsonMSFT committed:
Review feedback.

AaronRobinsonMSFT committed:
Add cleanup code to Finalizer.
Move main thread initialization/cleanup closer to when Main is called.

marek-safar commented on src/mono/System.Private.CoreLib/src/ILLink/ILLink.Descriptors.xml:
```diff
@@ -589,7 +589,11 @@
 		<!-- FIXME: -->
-		<type fullname="System.Threading.Thread" preserve="fields" />
+		<type fullname="System.Threading.Thread" preserve="fields">
+			<!-- threads.c -->
+			<method name="CreateAutoreleasePool" />
+			<method name="DrainAutoreleasePool" />
```
Hmm, this should be preferably in apple specific additional descriptor xml file

AaronRobinsonMSFT replied:
Thanks @marek-safar. I was hoping that was the solution.

AaronRobinsonMSFT committed:
Add Descriptors file for OSX.

AaronRobinsonMSFT committed:
Fix contract for new managed thread initialization from native.

jkotas commented on src/coreclr/vm/ceemain.cpp:
```diff
@@ -917,13 +918,13 @@ void EEStartupHelper()
 
         // throws on error
-        SetupThread();
+        mainThread = SetupMainThread();
```
This change is unnecessary now

AaronRobinsonMSFT committed:
Remove local mainThread.

AaronRobinsonMSFT committed:
Revert additional public API changes.

AaronRobinsonMSFT committed:
Revert comment changes.

jkotas commented on src/coreclr/vm/threads.h:
```diff
@@ -1764,23 +1777,31 @@ class Thread
     BOOL CreateNewThread(SIZE_T stackSize, LPTHREAD_START_ROUTINE start, void *args, LPCWSTR pName=NULL);
 
+private:
+    bool m_managedInitializationPerformed;
```
This flag does not seem to be really needed. The managed side has it duplicated.

AaronRobinsonMSFT replied:
This is for the specific native call not the managed logic. I appreciate that presently this call is restricted to only a managed call so could be handled on that end. My intent here is to avoid any consideration of making a managed call if the init hasn't been called. Given the minor size increase of the costly to activate `Thread` class I feel this is an acceptable trade off. Let me know if you think otherwise.

jkotas replied:
> My intent here is to avoid any consideration of making a managed call if the init hasn't been called.

Is this logic ever going to kick in? I believe that the managed call is always only made exactly once in with the current implementation.

AaronRobinsonMSFT replied:
>Is this logic ever going to kick in?

Yes since it isn't checked to avoid re-initialization but rather to permit always calling `Thread::CleanUpForManagedThreadInNative()`.

jkotas replied:
Neither of the two places that call `Thread::CleanUpForManagedThreadInNative` today need this check. Both of them call `Thread::CleanUpForManagedThreadInNative` only once the initialization has been done. The finalizer case has an extra line to guarantee that instead of always calling `Thread::CleanUpForManagedThreadInNative`.

AaronRobinsonMSFT replied:
Okay.

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Threading/Thread.OSX.cs:
```diff
@@ -0,0 +1,47 @@
+
+        internal static void CreateAutoreleasePool()
+        {
+            if (ThreadOSX.EnableAutoreleasePool)
+                s_AutoreleasePoolInstance = Interop.Sys.CreateAutoreleasePool();
```
```suggestion
            if (ThreadOSX.EnableAutoreleasePool)
            {
                Debug.Assert(s_AutoreleasePoolInstance == IntPtr.Zero);
                s_AutoreleasePoolInstance = Interop.Sys.CreateAutoreleasePool();
            }   
```

jkotas reviewed (approved):
LGTM otherwise. Thank you!

AaronRobinsonMSFT committed:
Review feedback.

AaronRobinsonMSFT commented on src/libraries/System.Private.CoreLib/src/System/Threading/Thread.OSX.cs:
```diff
@@ -0,0 +1,47 @@
+
+namespace System.Threading
+{
+    internal static class ThreadOSX
```
>I assume that you have done this intentionally to avoid adding static constructor to Thread.

Yes.

>Or maybe this can be just a part of ObjectiveCMarshal and be under FEATURE_OBJCMARSHAL ifdef. It is all one bundle of Objective-C supporting features that we are very likely to be enabled individually.

I am all for doing this after we merge together the two features. At present the `FEATURE_OBJCMARSHAL` is introduced in https://github.com/dotnet/runtime/pull/52146 which means until that goes in I can't reconcile that. Once this goes in, I will merge main into that branch and put all of it under that feature build define.

jkotas replied:
You give it the right names in this PR, and then the merge into the other branch will have smaller delta. I think it is better to avoid redundant renames in history, it makes it easier to see what happened. I do not care strong enough.

AaronRobinsonMSFT replied:
Fair point. Okay.

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Threading/AutoreleasePool.cs:
```diff
@@ -0,0 +1,48 @@
+            if (EnableAutoreleasePool
+                && s_AutoreleasePoolInstance != IntPtr.Zero)
+            {
+                Interop.Sys.DrainAutoreleasePool(s_AutoreleasePoolInstance);
```
```suggestion
                Interop.Sys.DrainAutoreleasePool(s_AutoreleasePoolInstance);
                s_AutoreleasePoolInstance = IntPtr.Zero;
```

AaronRobinsonMSFT committed:
Update the feature name to FEATURE_OBJCMARSHAL which is what it will
  eventually be when the Objective-C interop support is added.

AaronRobinsonMSFT committed:
Remove boolean member in Thread.

AaronRobinsonMSFT commented:
@lambdageek I've applied the feedback to Mono. Let me know what you think.

lambdageek reviewed (approved):
LGTM.  Thanks!
